core/src/main/java/io/micronaut/core/io/buffer/ReadBuffer.java

```diff
@@ -0,0 +1,169 @@
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.core.io.buffer;
+
+import io.micronaut.core.annotation.Internal;
+import io.micronaut.core.annotation.NonNull;
+
+import java.io.ByteArrayInputStream;
+import java.io.InputStream;
+import java.nio.charset.Charset;
+
+/**
+ * A buffer of bytes. Can be read from exactly once. Must be either consumed fully, or
+ * {@link #close() closed}.
+ *
+ * <h2 id="consuming">Consuming operations</h2>
+ *
+ * <p>Certain operations on ReadBuffers are <i>consuming</i>, meaning that they read all the
+ * remaining bytes in the buffer. Those operations are called out in the javadoc. After such an
+ * operation, the buffer does not need to be closed (though it's still allowed to do so).</p>
+ *
+ * @author Jonas Konrad
+ * @since 4.10.0
+ */
+public abstract class ReadBuffer implements AutoCloseable {
+    /**
+     * Internal constructor. Extension is only allowed in micronaut-core.
+     */
+    @Internal
+    protected ReadBuffer() {
+    }
+
+    /**
+     * Returns the number of bytes that can be read from this buffer.
+     *
+     * @return The number of readable bytes
+     */
+    public abstract int readable();
+
+    /**
+     * Create a new independent buffer that reads the same data as this one. This buffer remains
+     * unchanged.
+     *
+     * @return A new, independent buffer
+     * @throws IllegalStateException If this buffer is already closed or consumed
+     */
+    @NonNull
+    public abstract ReadBuffer duplicate();
+
+    /**
+     * Read some (but not all) bytes from this buffer. Those bytes will be removed from this
+     * buffer. This is <i>not</i> a <a href="#consuming">consuming operation</a>, even if
+     * {@code n == readable()}.
+     *
+     * @param n How many bytes to read
+     * @return A new, independent buffer that reads the first {@code n} bytes
+     * @throws IllegalStateException     If this buffer is already closed or consumed
+     * @throws IndexOutOfBoundsException If {@code n > readable()}
+     */
+    @NonNull
+    public abstract ReadBuffer readSome(int n) throws IndexOutOfBoundsException;
+
+    /**
+     * Create a new buffer with the same content as this one, consuming this buffer in the process.
+     * If this buffer is later closed, the returned buffer remains usable.
+     *
+     * <p>This is a <a href="#consuming">consuming operation</a>.
+     *
+     * @return a new ReadBuffer containing all the bytes that were previously in this ReadBuffer
+     * @throws IllegalStateException If this buffer is already closed or consumed
+     */
+    @NonNull
+    public abstract ReadBuffer move();
```

Comment:
It's not clear to me the purpose of this method. Consume and create a new buffer, why? Does it read all the data?

Reply:
we have this pattern in ByteBody already fwiw

Reply:
I don't understand why something needs to be moved

Reply:
Because `rb` is closed